### PR TITLE
Add preprocess function option to kernelshap

### PR DESCRIPTION
**Pull request by loostrum**
See #155

### CODE REVIEW

File: dianna/methods/kernelshap.py [hunk 1]
```diff
@@ -204,13 +206,13 @@ def _runner(self, features):
...
-            )
-        )[f"{self.output_node}"]
+        model_input = self._mask_image(features,
+                                       self.image_segments,
+                                       self.input_data,
+                                       self.background,
+                                       self.channels_axis_index,
+                                       self.input_node_dtype.as_numpy_dtype
+                                       )
+        if self.preprocess_function is not None:
```
Comment (geek-yang) [inline dianna/methods/kernelshap.py:216]: I wonder if it makes more sense to move the `preprocess_function` before calling the segmentation (line 122). If the user wants to preprocess the input data, I guess it happens before the segmentation and masking. What do you think @loostrum ?

**Review by geek-yang — APPROVED** [review]
Thanks for implementing the `preprocess_function`. Just a small comment about the location of this preprocess procedure.

File: dianna/methods/kernelshap.py [hunk 2]
```diff
@@ -147,6 +149,8 @@ def _prepare_image_data(self, input_data):
         Returns:
             transformed input data
         """
+        if self.preprocess_function is not None:
```
Comment (geek-yang) [inline dianna/methods/kernelshap.py:152]: I moved the `preprocess_function` to the function `_prepare_image_data`. There the `preprocess_function` is executed before the processing of input data inside `kernelshap` and therefore the user does not need to be bothered by the operations inside `kernelshap` when implementing their preferred way of preprocessing the input image. I think this is more friendly to the user.

**Review by geek-yang — APPROVED** [review]

**Comment by geek-yang** [comment]
And BTW, thanks for keeping the API consistent for all methods. I really love this kind of changes you made to our repo @loostrum!

File: dianna/methods/kernelshap.py (hunk 2 above)
Comment (loostrum) [reply dianna/methods/kernelshap.py:152]: I'm not sure I get why you moved it all the way to the start of the processing. What would then be the point of having a preprocessing function? The user could then just do it themselves before using dianna. 
In the other methods, the preprocess function is called as late as possible, just before running the model. The reason being that the model might require slightly different input than the XAI method. E.g. Images must have a channels axis, which you might not have for greyscale data. Then you could add the channels axis, and add a preprocessing function that removes the channels axis. Such a preprocessing  function should only be called just before running the model. 
You're absolutely right that it requires knowledge of the internals of the method, but I think that is exactly the point in this case

Comment (geek-yang) [reply dianna/methods/kernelshap.py:152]: Hi @loostrum , thanks for your explanation. My impression about `preprocess` comes from some other python packages (for instance, like [xarray.open_mfdataset](https://xarray.pydata.org/en/stable/generated/xarray.open_mfdataset.html). The process begins before the operations regarding the input. But I totally agree with you in this case and I think what you suggest here make more sense. Thanks again for the clarification and sorry for making this quite confusing. I will revert my change and just go ahead and merge it.

Comment (loostrum) [reply dianna/methods/kernelshap.py]: No problem, thanks for picking up this PR again!

**Pull request merged**